<a href="https://colab.research.google.com/github/abdulloyev/DataScientis-openSource/blob/main/Qandli_diabetni_bashorat_qilishning_intellektual_algoritmlari.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [73]:
import kagglehub
import pandas as pd
import os


In [74]:
# DataSetni yuklab olish
path = kagglehub.dataset_download("iammustafatz/diabetes-prediction-dataset")

print("DataSet fail manzili:", path)

DataSet fail manzili: /kaggle/input/diabetes-prediction-dataset


In [75]:
# fileni tekshirish
os.listdir(path)

['diabetes_prediction_dataset.csv']

In [76]:
# CSV faylni o‘qib olish
df = pd.read_csv(path + "/diabetes_prediction_dataset.csv")

# Dastlabki 5 ta qatorni ko‘rish
df.head(5)

gender   age  hypertension  heart_disease smoking_history    bmi  \
0  Female  80.0             0              1           never  25.19   
1  Female  54.0             0              0         No Info  27.32   
2    Male  28.0             0              0           never  27.32   
3  Female  36.0             0              0         current  23.45   
4    Male  76.0             1              1         current  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0

### `DataFrame` haqida ma'lumotlarni olamiz

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   gender               100000 non-null  object 
 1   age                  100000 non-null  float64
 2   hypertension         100000 non-null  int64  
 3   heart_disease        100000 non-null  int64  
 4   smoking_history      100000 non-null  object 
 5   bmi                  100000 non-null  float64
 6   HbA1c_level          100000 non-null  float64
 7   blood_glucose_level  100000 non-null  int64  
 8   diabetes             100000 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 6.9+ MB


### Qaysi ustunlarda ma'lumot yetishmayotganligi tekshirish

In [78]:
df.isnull().sum()

gender                 0
age                    0
hypertension           0
heart_disease          0
smoking_history        0
bmi                    0
HbA1c_level            0
blood_glucose_level    0
diabetes               0
dtype: int64

### `DataFrame` Harbir ustunlarni typelarni ko'rib olamiz

In [79]:
df.dtypes

gender                  object
age                    float64
hypertension             int64
heart_disease            int64
smoking_history         object
bmi                    float64
HbA1c_level            float64
blood_glucose_level      int64
diabetes                 int64
dtype: object

### Har bir ustun bo'yicha statistik ma'lumotlar (o'rtacha, min, max, standart chetlanish).

In [80]:
df.describe()

age  hypertension  heart_disease            bmi  \
count  100000.000000  100000.00000  100000.000000  100000.000000   
mean       41.885856       0.07485       0.039420      27.320767   
std        22.516840       0.26315       0.194593       6.636783   
min         0.080000       0.00000       0.000000      10.010000   
25%        24.000000       0.00000       0.000000      23.630000   
50%        43.000000       0.00000       0.000000      27.320000   
75%        60.000000       0.00000       0.000000      29.580000   
max        80.000000       1.00000       1.000000      95.690000   

         HbA1c_level  blood_glucose_level       diabetes  
count  100000.000000        100000.000000  100000.000000  
mean        5.527507           138.058060       0.085000  
std         1.070672            40.708136       0.278883  
min         3.500000            80.000000       0.000000  
25%         4.800000           100.000000       0.000000  
50%         5.800000           140.000000       0.000000  
75%         6.200000           159.000000       0.000000  
max         9.000000           300.000000       1.000000

Yuqoridagi natijaga etibor berib qaragan bo'lsak bizga keraksiz satrlar ham mavjud ekan ularni olib tashlashimiz kerak bo'ladi.

- bilamizki qandli diyabed ko'proq yoshi kattalarni bezovta qilib keladi, bizni dataFrameda yoshi `age<0` qiymatlar ham mavjud ekan bularni olib tashlaymiz

In [81]:
df = df[df['age'] > 18]

### `DataFrame` yetishmayotgan qiymatlarni to'ldirish

- biznig `dataframe` da `NaN` qiymatlar aniqlanmadi lekin buni ham o'rtacha qiymat bilan to'ldirishni ko'rib chiqamiz

In [82]:
df['hypertension'].fillna('Unknown', inplace=True)
df['age'].fillna(df['age'].median(), inplace=True)

<ipython-input-82-dfc10a55e297>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['hypertension'].fillna('Unknown', inplace=True)
<ipython-input-82-dfc10a55e297>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['hypertension'].fillna('Unknown', inplace=True)
<ipython-input-82-dfc10a55e297>:2: FutureWarning: A value is trying to be set on 

### Yangilangan ma'lumotlarni ko'rish

In [83]:
df.sample(5)

gender   age  hypertension  heart_disease smoking_history    bmi  \
44134    Male  39.0             0              0         No Info  33.62   
31844  Female  28.0             0              0          former  20.44   
81197  Female  22.0             0              0           never  22.46   
55276    Male  53.0             0              0         No Info  27.32   
48649  Female  52.0             0              0           never  29.48   

       HbA1c_level  blood_glucose_level  diabetes  
44134          6.5                   85         0  
31844          5.0                  130         0  
81197          5.8                  145         0  
55276          6.2                  155         0  
48649          3.5                  160         0

## Satrli ustunlarni `Encoder` yordamida raqamli ko'rinishga o'tgazib olamiz

In [84]:
df.dtypes

gender                  object
age                    float64
hypertension             int64
heart_disease            int64
smoking_history         object
bmi                    float64
HbA1c_level            float64
blood_glucose_level      int64
diabetes                 int64
dtype: object

In [85]:
from sklearn.preprocessing import LabelEncoder

# LabelEncoder obyektini yaratamiz
le = LabelEncoder()

# Kodlash kerak bo'lgan ustunlar
categorical_columns = ['gender', 'smoking_history']

# Har bir kategoriyal ustun uchun kodlash
for col in categorical_columns:
    df[col] = le.fit_transform(df[col])

<ipython-input-85-e6b4d0c28a4f>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = le.fit_transform(df[col])
<ipython-input-85-e6b4d0c28a4f>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = le.fit_transform(df[col])


### Yangilangan ma'lumotlarni ko'rish

In [86]:
df.sample(5)

gender   age  hypertension  heart_disease  smoking_history    bmi  \
75259       1  78.0             0              0                3  27.98   
13012       1  39.0             0              0                1  30.59   
85165       0  80.0             1              1                0  24.54   
84034       0  75.0             0              0                2  33.20   
42980       0  74.0             0              0                0  20.86   

       HbA1c_level  blood_glucose_level  diabetes  
75259          6.1                  145         0  
13012          4.0                  160         0  
85165          5.0                  159         0  
84034          6.1                  140         0  
42980          6.5                  126         0

## Modellar uchun kerak detallarni tayorlab olamiz

quyida diabetni bashorat qilish uchun 3 ta mashinani o‘rganish modeli — `Logistic Regression`, `K-Nearest Neighbors (KNN`), va `Random Forest` algoritmlarini qo‘llaymiz

In [87]:
# Kerakli kutubxonalar
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Modellar
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier


### `X` va `y` ni ajratamiz

In [88]:
X = df.drop("diabetes", axis=1)  # Kiruvchi belgilar
y = df["diabetes"]               # Chiqaruvchi belgi

### Train/Test bo‘lish va standartlashtirish

In [89]:
# Ma'lumotlarni bo‘lish
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# StandardScaler bilan normalizatsiya qilish
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


## Har bir modelni o‘rgatamiz va baholaymiz

## `📌 Logistic Regression`

In [90]:
log_model = LogisticRegression()
log_model.fit(X_train_scaled, y_train)
log_preds = log_model.predict(X_test_scaled)

print("Logistic Regression natijalari:")
print("Accuracy:", accuracy_score(y_test, log_preds))
print(classification_report(y_test, log_preds))


Logistic Regression natijalari:
Accuracy: 0.9525122549019608
              precision    recall  f1-score   support

           0       0.96      0.99      0.97     14639
           1       0.87      0.64      0.73      1681

    accuracy                           0.95     16320
   macro avg       0.91      0.81      0.85     16320
weighted avg       0.95      0.95      0.95     16320



## `📌 K-Nearest Neighbors (KNN)`

In [91]:
knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train_scaled, y_train)
knn_preds = knn_model.predict(X_test_scaled)

print("KNN natijalari:")
print("Accuracy:", accuracy_score(y_test, knn_preds))
print(classification_report(y_test, knn_preds))


KNN natijalari:
Accuracy: 0.9526348039215686
              precision    recall  f1-score   support

           0       0.96      0.99      0.97     14639
           1       0.88      0.62      0.73      1681

    accuracy                           0.95     16320
   macro avg       0.92      0.81      0.85     16320
weighted avg       0.95      0.95      0.95     16320



## `📌 Random Forest`

In [92]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
rf_preds = rf_model.predict(X_test)

print("Random Forest natijalari:")
print("Accuracy:", accuracy_score(y_test, rf_preds))
print(classification_report(y_test, rf_preds))

Random Forest natijalari:
Accuracy: 0.9638480392156863
              precision    recall  f1-score   support

           0       0.97      1.00      0.98     14639
           1       0.94      0.69      0.80      1681

    accuracy                           0.96     16320
   macro avg       0.95      0.84      0.89     16320
weighted avg       0.96      0.96      0.96     16320



## `✅ Model natijalarini solishtiramiz`

In [93]:
results = {
    "Model": ["Logistic Regression", "KNN", "Random Forest"],
    "Accuracy": [
        accuracy_score(y_test, log_preds),
        accuracy_score(y_test, knn_preds),
        accuracy_score(y_test, rf_preds)
    ]
}

results_df = pd.DataFrame(results)
results_df

Model  Accuracy
0  Logistic Regression  0.952512
1                  KNN  0.952635
2        Random Forest  0.963848

## **🥇 Eng yaxshi model: Random Forest**
**Random Forest** eng yuqori `accuracy = 0.963848` natijasini ko‘rsatdi. Bu model:

- Ko‘p daraxtli ansambl bo‘lib, har bir qaror daraxtining natijalarini birlashtiradi.

- Overfittingga nisbatan chidamli.

- Katta va murakkab datasetlarda yaxshi ishlaydi.

- Har bir xususiyatning (feature'ning) ahamiyatini ham aniqlay oladi.

## **🔍 Qisqacha tahlil:**

- **Logistic Regression:**	Soddaligi va tushunarliligi, ammo no-chiziqli bog'liqlikda zaif.

- **KNN:**	Mahalliy yaqinlikka asoslanadi, ammo katta datasetda sekin ishlaydi.

- **Random Forest:**	Yuqori aniqlik, barqarorlik, murakkab aloqalarni ham yaxshi o‘rganadi ✅